pkg/graveler/retention/starting_point_iterator.go

```diff
@@ -0,0 +1,84 @@
+package retention
+
+import "github.com/treeverse/lakefs/pkg/graveler"
+
+// A GCStartingPoint represents a commit from which the GC algorithm should start scanning.
+// It could be either a branch HEAD, or a dangling commit.
+type GCStartingPoint struct {
+	BranchID graveler.BranchID
+	CommitID graveler.CommitID
+}
+
+// GCStartingPointIterator combines a branch iterator and a commit iterator. Both are assumed to be sorted and to contain no duplicates.
+// Each returned GCStartingPoint object contains the commit id. If the entry came from the branch iterator, it also the branch id.
+// Commits appearing in both iterators appear only once and include the branch id.
+type GCStartingPointIterator struct {
+	commitIterator graveler.CommitIterator
+	branchIterator graveler.BranchIterator
+
+	commitValue *graveler.CommitRecord
+	branchValue *graveler.BranchRecord
+	value       *GCStartingPoint
+}
+
+func NewGCStartingPointIterator(commitIterator graveler.CommitIterator, branchIterator graveler.BranchIterator) *GCStartingPointIterator {
+	return &GCStartingPointIterator{commitIterator: commitIterator, branchIterator: branchIterator}
+}
+
+func (sp *GCStartingPointIterator) Next() bool {
+	prepareBranchValue := func() bool {
```

Comment:
think if you will change the order of the checks - we can remove this func too

Reply:
done